### DISCUSSION TITLE
ERROR: infinite symlink expansion detected in building jax from source in ubuntu?

### DISCUSSION THREAD

**Discussion by lzhhh93** [jax-ml/jax#8919] [Q&A]
I try to install jax and jaxlib from source, according to the official instruction(https://jax.readthedocs.io/en/latest/developer.html#building-from-source).

Cuda 11.5 is installed before and i have checked with nvidia-smi.

So to install them, i cloneed the source code and checkouted to the version that i want to install.
C++ compiler, Python packages numpy, six, wheel are installed.
Then i built jaxlib with CUDA support by running:

    python build/build.py --enable_cuda

it showed:

    ERROR: @local_config_cuda//:enable_cuda :: Error loading option 
    @local_config_cuda//:enable_cuda: Repository command failed
    Could not find any cudnn.h, cudnn_version.h matching version '' in any subdirectory:
        ''
        'include'
        'include/cuda'
        'include/*-linux-gnu'
        'extras/CUPTI/include'
            'include/cuda/CUPTI'
    of:
            '/lib'
            '/lib/i386-linux-gnu'
            '/lib/x86_64-linux-gnu'
            '/lib32'
            '/usr'
            '/usr/lib'
            '/usr/lib/i386-linux-gnu'
            '/usr/lib/x86_64-linux-gnu'
            '/usr/lib/x86_64-linux-gnu/libfakeroot'
            '/usr/lib32'
            '/usr/local/cuda'
            '/usr/local/cuda-11.1/targets/x86_64-linux/lib'
            '/usr/local/cuda/targets/x86_64-linux/lib'

so i copyed all the cudnn* files in /usr/local/cuda/include, according to this link https://stackoverflow.com/questions/31326015/how-to-verify-cudnn-installation.
And ran the same command again to install. Then it came up with this error:

    Analyzing: target //build:build_wheel (29 packages loaded, 117 targets configured)
    ERROR: infinite symlink expansion detected
    [start of symlink chain]
    /usr/bin/X11
    /usr/bin
    [end of symlink chain]
    ERROR: /home/control/.cache/bazel/_bazel_control/482c6a407c56a2bd5fbddc1ab25bd348/external/rules_cuda/cuda/BUILD:128:20: every rule of type cuda_toolchain_info implicitly depends upon the target '@local_cuda//:cuda/bin/nvcc', but this target could not be found because of: no such package '@local_cuda//': Symlink issue while evaluating globs: Infinite symlink expansion: /usr/bin/X11- > /usr/bin
    ERROR: Analysis of target '//build:build_wheel' failed; build aborted: Analysis failed
    INFO: Elapsed time: 18.105s
    INFO: 0 processes.
    FAILED: Build did NOT complete successfully (67 packages loaded, 612 targets configured)
    ERROR: Build failed. Not running target
    FAILED: Build did NOT complete successfully (67 packages loaded, 612 targets configured)
    b''

After i searched for solutions to this error, i found out that lots of solutions are coupled with tensorflow, which i haven't install and don't need for this project.

How should i deal with it? or at least where should i beginn with?

 
Btw: 
1. Jax version that i want to install is 0.1.68, and the project is not runnable with the newest version.
2. 'python' is pointed to 'python3.7' to run the command above to install, since i need jax to be installed in python3.7.

Many thanks.

**Comment by hawkinsp** [comment]
0.1.68 is 7 releases behind main, at this point. We most likely won't be able to help you if you cannot upgrade to the current release.

That said, I'd want to check: does `/usr/local/cuda` exist? It's usually a symlink: where does it point? If that's not where your CUDA installation is, where is it?

**Reply by lzhhh93** [reply]
control@control:/usr/local$ ls
bin   cuda-11    cuda-11.5  games    lib  sbin   src
cuda  cuda-11.1  etc        include  man  share

control@control:/usr/local/cuda$ ls -la
total 140
drwxr-xr-x 16 root root  4096 Dez  9 14:06 .
drwxr-xr-x 12 root root  4096 Dez  9 14:06 ..
drwxr-xr-x  3 root root  4096 Dez  9 14:06 bin
drwxr-xr-x  4 root root  4096 Dez  9 14:05 compute-sanitizer
drwxr-xr-x  3 root root  4096 Dez  9 14:05 doc
-rw-r--r--  1 root root   160 Nov 12 07:27 DOCS
-rw-r--r--  1 root root 61727 Nov 12 07:27 EULA.txt
drwxr-xr-x  5 root root  4096 Dez  9 14:06 extras
drwxr-xr-x  4 root root  4096 Dez  9 14:05 gds
lrwxrwxrwx  1 root root    28 Nov 18 19:12 include -> targets/x86_64-linux/include
lrwxrwxrwx  1 root root    24 Nov 15 19:51 lib64 -> targets/x86_64-linux/lib
drwxr-xr-x  7 root root  4096 Dez  9 14:05 libnvvp
drwxr-xr-x  2 root root  4096 Dez  9 14:05 nsightee_plugins
drwxr-xr-x  3 root root  4096 Dez  9 14:05 nvml
drwxr-xr-x  7 root root  4096 Dez  9 14:04 nvvm
-rw-r--r--  1 root root   524 Nov 12 07:27 README
drwxr-xr-x 11 root root  4096 Dez  9 14:06 samples
drwxr-xr-x  3 root root  4096 Dez  9 14:05 share
drwxr-xr-x  2 root root  4096 Dez  9 14:05 src
drwxr-xr-x  3 root root  4096 Dez  9 14:03 targets
drwxr-xr-x  2 root root  4096 Dez  9 14:06 tools
-rw-r--r--  1 root root  3036 Nov 19 05:35 version.json

control@control:/usr/local/cuda$ cd ..

control@control:/usr/local$ ls -la
total 48
drwxr-xr-x 12 root root 4096 Dez  9 14:06 .
drwxr-xr-x 12 root root 4096 Dez  4 19:40 ..
drwxr-xr-x  2 root root 4096 Dez 10 12:05 bin
lrwxrwxrwx  1 root root   22 Dez  9 14:06 cuda -> /etc/alternatives/cuda
lrwxrwxrwx  1 root root   25 Dez  9 14:06 cuda-11 -> /etc/alternatives/cuda-11
drwxr-xr-x 16 root root 4096 Dez 13 12:31 cuda-11.1
drwxr-xr-x 16 root root 4096 Dez  9 14:06 cuda-11.5
drwxr-xr-x  2 root root 4096 Sep 15 22:17 etc
drwxr-xr-x  2 root root 4096 Sep 15 22:17 games
drwxr-xr-x  4 root root 4096 Dez 10 16:44 include
drwxr-xr-x  5 root root 4096 Dez  8 12:08 lib
lrwxrwxrwx  1 root root    9 Dez  4 14:41 man -> share/man
drwxr-xr-x  2 root root 4096 Sep 15 22:17 sbin
drwxr-xr-x  7 root root 4096 Dez  8 20:03 share
